cluster-autoscaler/utils/drain/drain.go

```diff
@@ -74,188 +68,13 @@ const (
-			} else if err != nil {
-				return replicated, isDaemonSetPod, &BlockingPod{Pod: pod, Reason: UnexpectedError}, fmt.Errorf("error when trying to get daemonset for %s/%s , err: %v", pod.Namespace, pod.Name, err)
-			}
-		}
-	} else if refKind == "Job" {
-		if checkReferences {
-			job, err := listers.JobLister().Jobs(controllerNamespace).Get(controllerRef.Name)
-
-			// Assume the only reason for an error is because the Job is
-			// gone/missing, not for any other cause.  TODO(mml): something more
-			// sophisticated than this
-			if err == nil && job != nil {
-				replicated = true
-			} else {
-				return replicated, isDaemonSetPod, &BlockingPod{Pod: pod, Reason: ControllerNotFound}, fmt.Errorf("job for %s/%s is not available: err: %v", pod.Namespace, pod.Name, err)
-			}
-		} else {
-			replicated = true
-		}
-	} else if refKind == "ReplicaSet" {
-		if checkReferences {
-			rs, err := listers.ReplicaSetLister().ReplicaSets(controllerNamespace).Get(controllerRef.Name)
-
-			// Assume the only reason for an error is because the RS is
-			// gone/missing, not for any other cause.  TODO(mml): something more
-			// sophisticated than this
-			if err == nil && rs != nil {
-				if rs.Spec.Replicas != nil && *rs.Spec.Replicas < minReplica {
-					return replicated, isDaemonSetPod, &BlockingPod{Pod: pod, Reason: MinReplicasReached}, fmt.Errorf("replication controller for %s/%s has too few replicas spec: %d min: %d",
-						pod.Namespace, pod.Name, rs.Spec.Replicas, minReplica)
-				}
-				replicated = true
-			} else {
-				return replicated, isDaemonSetPod, &BlockingPod{Pod: pod, Reason: ControllerNotFound}, fmt.Errorf("replication controller for %s/%s is not available, err: %v", pod.Namespace, pod.Name, err)
-			}
-		} else {
-			replicated = true
-		}
-	} else if refKind == "StatefulSet" {
-		if checkReferences {
-			ss, err := listers.StatefulSetLister().StatefulSets(controllerNamespace).Get(controllerRef.Name)
-
-			// Assume the only reason for an error is because the StatefulSet is
-			// gone/missing, not for any other cause.  TODO(mml): something more
-			// sophisticated than this
-			if err == nil && ss != nil {
-				replicated = true
-			} else {
-				return replicated, isDaemonSetPod, &BlockingPod{Pod: pod, Reason: ControllerNotFound}, fmt.Errorf("statefulset for %s/%s is not available: err: %v", pod.Namespace, pod.Name, err)
-			}
-		} else {
-			replicated = true
-		}
-	}
-
-	return replicated, isDaemonSetPod, &BlockingPod{}, nil
-}
-
 // ControllerRef returns the OwnerReference to pod's controller.
 func ControllerRef(pod *apiv1.Pod) *metav1.OwnerReference {
 	return metav1.GetControllerOf(pod)
 }
 
-// isPodTerminal checks whether the pod is in a terminal state.
-func isPodTerminal(pod *apiv1.Pod) bool {
+// IsPodTerminal checks whether the pod is in a terminal state.
+func IsPodTerminal(pod *apiv1.Pod) bool {
```

Comment:
Any reason not to move `IsPodTerminal`, `HasSafeToEvictAnnotation`, etc. to the respective rules?

Reply:
I've been treating this as a public library and trying to reduce disruption to downstream dependencies. If such changes are okay, I can be more aggressive in my refactoring.

I would like to do this in a follow-up PR if that's okay.

Reply:
Sure, let's leave it out of scope for this one, it's already quite large.